aws-api-appsync/src/main/java/com/amplifyframework/datastore/appsync/ModelWithMetadataAdapter.java

```diff
@@ -58,4 +64,24 @@ public static void register(@NonNull GsonBuilder builder) {
         ModelMetadata metadata = context.deserialize(json, ModelMetadata.class);
         return new ModelWithMetadata<>(model, metadata);
     }
+
+    @Override
+    public JsonElement serialize(
+            ModelWithMetadata<? extends Model> src, Type typeOfSrc, JsonSerializationContext context) {
+        JsonObject result = new JsonObject();
+
+        // Flatten out the fields of the model and its metadata into a flat key-value map.
+        // To do this, serialize each individually, and then add the key/value pairs for each
+        // object into a new container.
+        JsonObject serializedMetadata = (JsonObject) context.serialize(src.getSyncMetadata());
+        for (Map.Entry<java.lang.String, JsonElement> entry : serializedMetadata.entrySet()) {
+            result.add(entry.getKey(), entry.getValue());
+        }
+        JsonObject serializedModel = (JsonObject) context.serialize(src.getModel());
+        for (Map.Entry<java.lang.String, JsonElement> entry : serializedModel.entrySet()) {
+            result.add(entry.getKey(), entry.getValue());
```

Comment:
There's the possibility here for a user key to conflict with a metadata key. Are we making it clear to the user somewhere what these reserved metadata keys are so they don't accidentally make a property with the same name?

Reply:
This is true! But, I think the conflict happens earlier in the system, than this point. By this point, we're already operating on a model that was generated by the CLI. So, the conflict would have been caught there.

I did a quick audit of where/how these fields are mentioned, and I agree it could use improvement.

AppSync & the CLI automatically include a range of fields. Some are prefixed with `_` (e.g., `_lastChangedAt`, `_deleted`, `_version`), while others are not (`updatedAt`, `createdAt`.)

We have a few pieces of existing documentation about this:

In the [GraphQL Directive usage](https://docs.amplify.aws/cli/graphql-transformer/directives#usage):

> Model directive automatically adds `createdAt` and `updatedAt` timestamps to each entities.

`_lastChangedAt`, `_version`, and `_deleted` are mentioned in a few places in the documentation, however, not explicitly as reserved words in the GraphQL schema that the user provides.
1. [DataStore "How it works" Guide to using the AppSync Console](https://docs.amplify.aws/lib/datastore/how-it-works/q/platform/android#writing-data-from-the-appsync-console)
2. [AppSync schema requirements for using versioned versioned data](https://docs.aws.amazon.com/appsync/latest/devguide/conflict-detection-and-sync.html#versioned-data-source-metadata)